flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/CancelingTest.java

```diff
@@ -0,0 +1,171 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.declarative;
+
+import org.apache.flink.api.common.JobStatus;
+import org.apache.flink.runtime.JobException;
+import org.apache.flink.runtime.client.JobExecutionException;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.executiongraph.ExecutionGraph;
+import org.apache.flink.runtime.executiongraph.TaskExecutionStateTransition;
+import org.apache.flink.runtime.executiongraph.TestingExecutionGraphBuilder;
+import org.apache.flink.runtime.scheduler.ExecutionGraphHandler;
+import org.apache.flink.runtime.scheduler.InternalFailuresListener;
+import org.apache.flink.runtime.scheduler.OperatorCoordinatorHandler;
+import org.apache.flink.runtime.taskmanager.TaskExecutionState;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+/** Tests for the {@link Canceling} state of the declarative scheduler. */
+public class CancelingTest extends TestLogger {
+
+    @Test
+    public void testExecutionGraphCancelationOnEnter() throws Exception {
+        try (MockStateWithExecutionGraphContext ctx = new MockStateWithExecutionGraphContext()) {
+            RestartingTest.CancellableExecutionGraph cancellableExecutionGraph =
+                    new RestartingTest.CancellableExecutionGraph();
+            Canceling canceling = createCancelingState(ctx, cancellableExecutionGraph);
+
+            canceling.onEnter();
+            assertThat(cancellableExecutionGraph.isCancelled(), is(true));
+        }
+    }
+
+    @Test
+    public void testTransitionToFinishedWhenCancellationCompletes() throws Exception {
+        try (MockStateWithExecutionGraphContext ctx = new MockStateWithExecutionGraphContext()) {
+            Canceling canceling = createCancelingState(ctx);
+
+            ctx.setExpectFinished(
+                    archivedExecutionGraph ->
+                            assertThat(archivedExecutionGraph.getState(), is(JobStatus.CANCELED)));
+            canceling.getExecutionGraph().cancel(); // this calls onTerminalState.onTerminalState()
+        }
+    }
+
+    @Test
+    public void testTransitionToSuspend() throws Exception {
+        try (MockStateWithExecutionGraphContext ctx = new MockStateWithExecutionGraphContext()) {
+            Canceling canceling = createCancelingState(ctx);
+
```

Comment:
I assume that you don't call `onEnter` because it would make the state go to `Finished`. I think that we should stick to the convention of `State` and call `onEnter`. Otherwise we risk that we are testing something which is impossible to reach in production. The proper solution would be to ensure that the `ExecutionGraph` does not directly go to `CANCELLED` when `onEnter` is called.